apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc/states.rs

```diff
@@ -0,0 +1,168 @@
+//! All the states of the [`BlockSpaceAllocator`] state machine,
+//! over the extent of a Tendermint consensus round
+//! block proposal.
+//!
+//! # States
+//!
+//! The state machine moves through the following state tree:
+//!
+//! 1. [`BuildingDecryptedTxBatch`] - the initial state. In
+//!    this state, we populate a block with DKG decrypted txs.
+//! 2. [`BuildingProtocolTxBatch`] - the second state. In
+//!    this state, we populate a block with protocol txs.
+//! 3. [`BuildingEncryptedTxBatch`] - the third state. In
+//!    this state, we populate a block with DKG encrypted txs.
+//!    This state supports two modes of operation, which you can
+//!    think of as two states diverging from [`BuildingProtocolTxBatch`]:
+//!   * [`WithoutEncryptedTxs`] - When this mode is active, no encrypted txs are
+//!     included in a block proposal.
+//!   * [`WithEncryptedTxs`] - When this mode is active, we are able to include
+//!     encrypted txs in a block proposal.
+//! 4. [`FillingRemainingSpace`] - the fourth and final state.
```

Comment:
We shouldn't include arbitrary transactions, as per our previous discussions. If we put in encrypted txs, we could accidentally fill more then 1/3 of the block with encrypted txs. We should only take extra protocol txs. And if there is space left, that's life.

Reply:
ah, this is being enforced in #774 - https://github.com/anoma/namada/blob/fb92431636a2f4ecbc22ef5e33992c300c7e4fbc/apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc/states/remaining_txs.rs#L27-L47

Reply:
albeit now that I think about, there's still a bug here. we could be violating the 1/3 ratio

Reply:
alright fair enough, we'll limit the last state to only include new protocol txs